pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"net"
+	"testing"
+)
+
+func TestAllIPBlocksAvailable(t *testing.T) {
+	ips := [8]string{"192.168.92.32/29", "192.168.92.40/29", "192.168.92.48/29", "192.168.92.56/29"}
+
+	cidr := "192.168.92.32/27"
+
+	ip, err := GetUnreservedIPBlock(make(map[string]bool), cidr)
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+	if ip != ips[0] {
+		t.Errorf("Expected IP %s to be released from the free IP pool but got IP %s", ips[0], ip)
+	}
+}
+
+func TestSomeIPBlocksAvailable(t *testing.T) {
+	ips := [8]string{"192.168.92.32/29", "192.168.92.40/29", "192.168.92.48/29", "192.168.92.56/29"}
+
+	cidr := "192.168.92.32/27"
+	reservedIPBlocks := make(map[string]bool)
+	for i := 0; i < 2; i++ {
+		reservedIPBlocks[ips[i]] = true
+	}
+
+	ip, err := GetUnreservedIPBlock(reservedIPBlocks, cidr)
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+	if ip != ips[2] {
+		t.Errorf("Expected IP %s to be released from the free IP pool but got IP %s", ips[2], ip)
+	}
+}
+
+func TestNoIPBlocksAvailable(t *testing.T) {
+	ips := [8]string{"192.168.92.33/29", "192.168.92.42/29", "192.168.92.49/29", "192.168.92.58/29"}
+
+	cidr := "192.168.92.32/27"
+	reservedIPBlocks := make(map[string]bool)
+	for _, ip := range ips {
+		reservedIPBlocks[ip] = true
+	}
+
+	ip, err := GetUnreservedIPBlock(reservedIPBlocks, cidr)
+	if err == nil {
+		t.Errorf("Expected error as all IPs in cidr %s had been reserved  but got IP %s as unreserved", cidr, ip)
+	}
+}
+
+func TestValidateCIDR(t *testing.T) {
+	cidr1 := "192.168.92.32/27"
+	cidr2 := "192.168.92.48/26"
+
+	err := validateCIDROverlap(cidr1, cidr2)
+	if err == nil {
+		t.Errorf("Expected error as cidr %s overlaps with cidr %s", cidr1, cidr2)
+	}
+
+	cidr2 = "192.168.22.67/26"
+	err = validateCIDROverlap(cidr1, cidr2)
+	if err != nil {
+		t.Errorf("Got overlapping error for non overlapping cidrs %s and %s", cidr1, cidr2)
+	}
+
+}
+
+func TestIncrementIP(t *testing.T) {
```

Comment:
Table driven tests. This one is not comprehensive. Try adding some edge cases and funky input. Error cases are also good